pkg/controller/operators/olm/operator.go

```diff
@@ -158,6 +187,80 @@ func (a *Operator) syncDeployment(obj interface{}) (syncError error) {
 	return nil
 }
 
+func (a *Operator) updateDeploymentAnnotation(op *v1alpha2.OperatorGroup) error {
+	selector, err := metav1.LabelSelectorAsSelector(&op.Spec.Selector)
+	if err != nil {
+		return err
+	}
+	operatorGroupOpts := metav1.ListOptions{LabelSelector: selector.String()}
+	// TODO(jpeeler): this needs to use user impersonation with op.Spec.ServiceAccount
+	namespaceList, err := a.OpClient.KubernetesInterface().CoreV1().Namespaces().List(operatorGroupOpts)
+	if err != nil {
+		return err
+	}
+
+	nsCount := len(namespaceList.Items)
+
+	if len(op.Status.Namespaces) == nsCount {
+		for i, v := range namespaceList.Items {
+			if v.Name != op.Status.Namespaces[i].Name {
+				break
+			}
+		}
+		// status is current with correct namespaces, so no further updates required
+		return nil
+	}
+	op.Status.Namespaces = namespaceList.Items
+	op.Status.LastUpdated = timeNow()
+
+	// make string list made up of comma separated namespaces
+	var nsList strings.Builder
+	for i := 0; i < nsCount-1; i++ {
+		nsList.WriteString(namespaceList.Items[i].Name + ",")
+	}
+	nsList.WriteString(namespaceList.Items[nsCount-1].Name)
+
+	// write namespaces to watch in every deployment
+	currentNamespace := op.GetNamespace()
+	csvsInNamespace := a.csvsInNamespace(currentNamespace)
+	for csvName, csv := range csvsInNamespace {
+		strategy, err := a.resolver.UnmarshalStrategy(csv.Spec.InstallStrategy)
+		if err != nil {
+			return fmt.Errorf("error unmarshaling strategy from ClusterServiceVersion '%s' with error: %s", csvName, err)
+		}
+
+		strategyDetailsDeployment, ok := strategy.(*install.StrategyDetailsDeployment)
+		if !ok {
+			return fmt.Errorf("could not assert strategy implementation as deployment for CSV %s", csvName)
+		}
+
+		for _, deploy := range strategyDetailsDeployment.DeploymentSpecs {
+			deploy.Spec.Template.Annotations["olm.targetNamespaces"] = nsList.String()
+			_, err := a.client.Operators().ClusterServiceVersions(currentNamespace).Update(csv)
+			if err != nil {
+				return fmt.Errorf("CSV update for '%v' failed: %v\n", csvName, err)
+			}
+			a.requeueCSV(csvName, currentNamespace)
```

Comment:
this will requeue the same csv multiple times if there're multiple deployments in the same csv

I'm also not sure we need to requeue - if this update causes the csv deployment to go unhealthy, we should detect that as an issue in the CSV anyway.